File: src/main/java/org/kite9/diagram/adl/DiagramElement.java
package org.kite9.diagram.adl;

import org.kite9.diagram.common.HintMap;
import org.kite9.diagram.position.Direction;
import org.kite9.diagram.position.RenderingInformation;

/**
 * Parent class for all elements of the diagram.
 */
public interface DiagramElement extends Comparable<DiagramElement> {

	public DiagramElement getParent();
	
	/**
	 * ID should be a project-unique ID to describe this element.  It is also used within the 
	 * XML to allow references between the elements of the XML file. 
	 * 
	 * ID is also used for hashcode and equals.  Set an ID to ensure sorting, maps
	 * and therefore diagram layouts, are deterministic.
	 * 
	 * IDs are expected for most elements, but are optional.
	 * 
	 */
	public String getID();

	public RenderingInformation getRenderingInformation();
	
	@Deprecated
	public void setRenderingInformation(RenderingInformation ri);
	
	public HintMap getPositioningHints();
		
	@Deprecated
	public String getShapeName();
	
	/**
	 * This method returns the container.  Since in the future, not all `Connected` objects will be immediately 
	 * in containers, we should have this deprecated.
	 * @return
	 */
	@Deprecated
	Container getContainer();
	
	public double getMargin(Direction d);
		
}

File: a/src/myMain/kotlin/org/kite9/diagram/model/DiagramElement.kt
package org.kite9.diagram.model

import org.kite9.diagram.common.algorithms.det.Deterministic
import org.kite9.diagram.model.position.RenderingInformation

/**
 * Parent class for all elements of the diagram.
 */
interface DiagramElement : Comparable<DiagramElement>, Deterministic {

    /**
     * Returns the parent element, or null if there is no parent.
     */
    fun getParent(): DiagramElement?

    fun getRenderingInformation(): RenderingInformation

    /**
     * Returns the number of levels deep which this element is embedded in the diagram hierarchy, with zero the top level.
     */
    fun getDepth(): Int

    /**
     * Similar to parent element, but sometimes the parent doesn't do the containing.  e.g. with a terminator, the
     * container will be the thing the terminator is in, whereas the parent will be the link.
     */
    fun getContainer() : Container?
}